data-pipeline-flink/assessment-aggregator/src/main/scala/org/sunbird/dp/assessment/functions/AssessmentAggWindowFun.scala

```diff
@@ -0,0 +1,197 @@
+      .value("user_id", assessment.userId)
+      .value("content_id", assessment.contentId)
+      .value("attempt_id", assessment.attemptId)
+      .value("updated_on", new DateTime(DateTimeZone.UTC).getMillis)
+      .value("created_on", assessment.aggregate.createdOn.get)
+      .value("last_attempted_on", assessment.aggregate.lastAttemptedOn.get)
+      .value("total_score", assessment.aggregate.totalScore)
+      .value("total_max_score", assessment.aggregate.totalMaxScore)
+      .value("question", assessment.aggregate.questionsList.asJava)
+      .value("grand_total", assessment.aggregate.grandTotal)
+  }
+
+  def insertData(batchWriteSize: Int, queriesList: List[Insert])(implicit metrics: Metrics): Unit = {
+    val groupedQueries = queriesList.grouped(batchWriteSize).toList
+    groupedQueries.foreach(queries => {
+      val cqlBatch = QueryBuilder.batch()
+      queries.map(query => cqlBatch.add(query))
+      val result = cassandraUtil.upsert(cqlBatch.toString)
+      if (result) {
+        metrics.incCounter(config.dbUpdateCount)
+      } else {
+        metrics.incCounter(config.dbUpdateFailed)
+        val msg = "Database update has failed: " + cqlBatch.toString
+        logger.error(msg)
+        throw new Exception(msg)
+      }
+    })
+  }
+
+  def readAssessment(event: Event): Row = {
+    val query = QueryBuilder.select("last_attempted_on", "created_on")
+      .from(config.dbKeyspace, config.dbTable).
+      where(QueryBuilder.eq("attempt_id", event.attemptId))
+      .and(QueryBuilder.eq("batch_id", event.batchId))
+      .and(QueryBuilder.eq("user_id", event.userId))
+      .and(QueryBuilder.eq("content_id", event.contentId))
+      .and(QueryBuilder.eq("course_id", event.courseId)).toString
+    cassandraUtil.findOne(query)
+  }
+
+  def isValidContent(courseId: String, contentId: String)(metrics: Metrics): Boolean = {
+    val leafNodes = dataCache.getKeyMembers(key = s"$courseId:$courseId:leafnodes")
+    if (!leafNodes.isEmpty) {
+      metrics.incCounter(config.cacheHitCount)
+      leafNodes.contains(contentId)
+    } else {
+      metrics.incCounter(config.cacheHitMissCount)
+      true
+    }
+  }
+
+  def getQuestion(questionData: QuestionData, assessTs: Long): UDTValue = {
+
+    questionType.newValue().setString("id", questionData.item.id).setDouble("max_score", questionData.item.maxscore)
+      .setDouble("score", questionData.score)
+      .setString("type", questionData.item.`type`)
+      .setString("title", questionData.item.title)
+      .setList("resvalues", getListValues(questionData.resvalues).asJava).setList("params", getListValues(questionData.item.params).asJava)
+      .setString("description", questionData.item.desc)
+      .setDecimal("duration", BigDecimal.valueOf(questionData.duration)).setTimestamp("assess_ts", new Timestamp(assessTs))
+  }
+
+  def getListValues(values: util.List[util.HashMap[String, Any]]): mutable.Buffer[util.Map[String, Any]] = {
+    values.asScala.map { res =>
```

Comment:
Please use Scala lists everywhere. It looks like you want to return List[Map[String, String]] but the return type is List[Map[String, Any]]. What is the purpose of this JSON serialization?